### ISSUE TITLE
What was it like to learn Elixir?

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/elixir#504]
We’ve recently [started a project](https://exercism.io/blog/redesigning-tracks-in-partnership-with-chicago-university-and-sloan-foundation) to find the best way to design our tracks, in order to optimize the learning experience of students.

As a first step, we’ll be [examining the ways in which languages are unique and the ways in which they are similar](https://github.com/exercism/exercism/issues/5059). For this, we’d really like to use the knowledge of **everyone** involved in the Exercism community (students, mentors, maintainers) to answer the following questions:

1. How was your experience learning Elixir? What was helpful while learning Elixir? What did you struggle with? How did you tackle problems?
1. In what ways did Elixir differ from other languages you knew at the time? What was hard to learn? What did you have to unlearn? What syntax did you have to remap? What concepts carried over nicely?

Could you spare 5 minutes to help us by answering these questions? It would greatly help us improve the experience students have learning Elixir :)

_Note: this issue is **not** meant as a discussion, just as a place for people to post their own, personal experiences._

_Want to keep your thoughts private but still help? Feel free to email me at erik@exercism.io_

Thank you!

**Comment by neenjaw** [comment]
My background, I completed my computer science degree in 2006 (4 months after jQuery was released, to put in the perspective of web development) and always had an interest in web programming, web applications.  Then life happened and somehow wound up with a second degree in nursing (it still feels like an odd transition at times, and maybe someday it will transition back, but for now it keeps the lights on), so it was relegated to hobby work in which elixir piqued my interest.

## 1

At first, I would say learning elixir was hard to get started for a few reasons:
- I was coming from more recent experience in imperative- and OO-based languages (PHP, JavaScript, Java)
- I had no experience with ruby

Once I had a basic grasp of the syntax and core concepts with the elixir-lang.org's getting started guide, and an e-book I'd picked up in a humble bundle deal, I wanted to learn how to use it well, so started down the exercism elixir track, opting for the mentored experience.

I don't think I struggled conceptually with the language having been exposed to Haskell (functional language giving insight to immutability and function purity) and prolog (pattern matching).

I set out to tackle problems be working to reduce them to their parts so that i could then compose them into a system.  And what I like about elixir is that I don't have lots of boilerplate that I have to type and memorize to accomplish basic functions and logical composition.

The biggest problems I struggled with were the meta-programming exercises as that was a completely new topic for me.  But the language documentation is great with a rich standard library, the erlang documentation is rich (though not always novice-friendly), elixir-school has at least a little on almost every topic to get you started. and `elixirforum` was a great resource of code examples.

I did find that googling randomly would often pull up pretty old content from stack overflow (code from before version 1.6 -- which is considered the stable versions from then on)

One of the biggest helps I think I had was just reading lots of code in the community solutions, which I continue to try to promote to students.

## 2

Elixir differed in its syntax, concurrency paradigm (agent based) from other languages.  Playing around with prolog made transitioning to elixir fairly easy since the functional concepts are very similar for list processing (`foldl` in prolog became `Enum.reduce` in elixir) and composing logic.

I found elixir very easy to compartmentalize from other languages (but wish that I could now take elixir to other languages -- e.g. I really wish javascript had a pipe operator `|>`), so I don't think I had to unlearn any thing.

**Comment by devonestes** [comment]
Funny enough, for me it was really easy. I'm always a little concerned by that because I know most people say it's kind of hard, but I don't have a CS degree (although I already had a good understanding of recursion and many FP principles).

The syntax was totally natural for me, and I actually find programming with functions and immutable data far easier to understand than I do OOP despite having used it for many years.

After going through the introduction documentation on the Elixir website and doing those exercises, I was basically there. After doing that I did the track on Exercism, and I was already at a point where I felt comfortable with like 90% of the language and its features. Luckily there aren't actually a whole ton of things that Elixir does, and it's an extremely consistent language, so there weren't many of those strange corner cases in the standard library that don't work like you'd expect and you just need to remember. The documentation is also great and easily searchable, so while I didn't know all the APIs at first, I knew where I could find them and learning them was easy.

Elixir is different from other languages in that it has OTP built into the language and the standard library. Supervision strategies are easy to understand, but hard to apply in practice as it requires real thought about the behavior you want your system to have. Of course anything distributed comes with a whole host of unique and difficult problems, but that's not exclusive to Elixir.

**Comment by angelikatyborska** [comment]
## My background

- 2012: I did an advanced class in CS in high school where I first saw a programming language - Pascal and Delphi.

- 2013-2014: I was studying a technical subject, but not CS, at Uni, where I learned Matlab. I was teaching myself front-end web development in my free time, got to know JavaScript and jQuery.

- 2014-2015: I was studying CS (dropped out during the 3rd semester). I got to know C/C++, Java, Prolog, some basic knowledge about operating systems. I also taught myself some PHP and Ruby outside of Uni. I fell in love with Ruby and wanted a Ruby job ASAP ;).

- 2016: I was a junior Ruby dev, and the seniors made a decision to do new projects in Elixir, so I had to learn it.


## 1. How was your experience learning Elixir? What was helpful while learning Elixir? What did you struggle with? How did you tackle problems?

At work, we got a presentation from a senior dev, an introduction to Elixir. Then I read the whole "Getting Started" guide on elixir-lang.org. It was extremely helpful and it was enough to start writing Elixir at work.

I couldn't understand charlists at first. I didn't know that chars are integers, I didn't know how to write a char, only a charlist. I was very confused the first time I wrote a list of numbers that happened to be printable char codepoints and it was displayed as a charlist, not a list of numbers. It took me months before I found out this syntax: `?A`.

I struggled a lot with reading Erlang code. We did a web app with _just_ cowboy, no Phoenix, no Plug. I couldn't understand the code samples in Erlang. I didn't know Erlang then, it's only slightly better now. I don't have to deal with Erlang often, so I am not very motivated to learn it well. I tried to do Erlang on Exercism but I don't enjoy it very much. This is still an unresolved issue for me.

I also struggled with OTP. I read the guides multiple times. I was confused by the names. When I hear "GenServer", I ask myself "what is it generating?!". I wish it had the full name "GenericServer". I would always forget the difference between `handle_call`, `handle_cast`, and `handle_info`. I kinda understand "call" (like you call a function and get a value back), but "cast" and "info"? I only know the term "casting" in the sense of automatically changing the type of some data, like casting an int to a float. What does it have to do with "sends an asynchronous request to the server"? Frankly, I still haven't found any source that would make me feel like I really understand and I can use OTP. But I am not actively looking because it hasn't been a big issue at work so far.

The biggest help with Elixir for me were the two senior devs I worked with at my first job. Being a junior, I found it hard to even put in words what I don't understand, so sometimes searching online was not an option, but having another human being look at my screen was perfect.

Elixir's excellent documentation, and the fact that documentation is a first-class citizen available in iex, was also a huge help.

## 2. In what ways did Elixir differ from other languages you knew at the time? What was hard to learn? What did you have to unlearn? What syntax did you have to remap? What concepts carried over nicely?

That one Prolog class I had in Uni was the best thing that happened for my understanding of Elixir. We basically went, step-by-step, through exercises like Exercism's ListOps, and then way more complicated ones. The teacher explained pattern-matching, recursion, accumulators, tail calls. It was all very easy to me when I came to Elixir.

It was hard however, to switch from OOP thinking to FP thinking. I have heard about design patterns, I read a few books about good practices, and none of that applied to FP. I had to do FP without ever seeing examples of an app well-written using FP.

Elixir was way more explicit than Ruby. I hated that for the first few months because it required more "unnecessary" typing. I learned to love it when I had to jump into a new existing project and realized how much easier it is thanks to that explicitness.

I found it very hard to accept that there are no arrays in Elixir. I was using lists as I would use arrays in other languages, until another Elixir mentor explained to me why I shouldn't (thanks!).

**Comment by OvermindDL1** [comment]
I've programmed my whole life as well, learning Elixir was just another layer on Erlang, which I'd already have known for about 15 years prior to Elixir.  I mostly like Elixir for it's macro capabilities, that and mix was the big boons of it over Erlang, otherwise I prefer Erlang's syntax as it is significantly less noisy.  It was a pretty trivial learning though since it is just erlang with a more noisy syntax.

**Comment by nathanielknight** [comment]
Not sure if this is still helpful, but my experience as a learner:

# 1

So far, learning Elixir has been pretty smooth for me. I've used functional and
immutable programming languages in the past (Scheme, Clojure, Elm) so many of the
concetps were familiar. I do struggle to make good use of the many functions in the
standard library; Exercism is a boon in this regard, as folks are keen to point out
where standard tools can be applied. I've also been tripped up by some of the ways
Elixir uses interfaces, which are powerful but not obvious. Striving to use things like
`Access` in exercises is how I'm trying to get comfortable with these.

All that said, I've yet to get to grips with building anything real with Elixir, and OTP
/ process management, packaging and deployment are still ahead of me, so my experience
will likely change in the future.


# 2

In what ways did Elixir differ from other languages you knew at the time? What was hard
to learn? What did you have to unlearn? What syntax did you have to remap? What concepts
carried over nicely?

The big things that makes Elixir different from the other languages I know is
it's process model, which I've yet to get to grips with.

In terms of syntax and more tactical code construction, building functions that work
together well (e.g. by chaining nicely with pipes) and make good use of the standard
library is what feels like a challenge. Since my professional workhorse langauge is
Python, I tend to assign more variables than is necessary, and I often want to reach for
mutation to solve problems, which is a habit that's difficult to un-learn.

That said, having used immutable data structures before (e.g. while reading *Structure
and Interpretation of Computer Programs* or learning Clojure and Elm) was very, very
helpful for getting comfortable with Elixir.

**Comment by Cohen-Carlisle** [comment]
Better late than never...

> 1. How was your experience learning Elixir? What was helpful while learning Elixir? What did you struggle with? How did you tackle problems?

My experience was pretty smooth sailing in that I was a ruby dev that was attracted to functional programming techniques and wished ruby had _less_ ways to do things. I mostly learned elixir by doing exercism v1 problems and viewing community solutions, along with reading the docs. I did struggle with fully understanding recursion and dealing with immutability until the List Ops exercise on exercism. I approached learning in a very bottom-up way where I wanted to understand each piece well as I put them together. I also was very motivated to produce highly idiomatic and "clean" solutions.

> 2. In what ways did Elixir differ from other languages you knew at the time? What was hard to learn? What did you have to unlearn? What syntax did you have to remap? What concepts carried over nicely?

While the syntax is very similar to ruby, as I mentioned, I did struggle with really understanding recursion and dealing with immutability until I did List Ops. I definitely had to unlearn mutation and had to learn to be more disciplined about using returned values instead. Still, between elixir's docs and experimenting more with ruby's functional methods, it was relatively smooth sailing.

All that said, I still don't think I really understand OTP and how to build complex systems with it. I think OTP is also a hard thing to do on exercism as exercism (in my experience) is more about solving toy problems to practice idiomatic approaches in your language thank building scalable systems.
